src/nupic/regions/BacktrackingTMRegion.hpp

```diff
@@ -0,0 +1,157 @@
+/* ---------------------------------------------------------------------
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2018, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ *
+ * Author: David Keeney, June, 2018  (ported from Python)
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+ * Declarations for TMRegion class
+ */
+
+//----------------------------------------------------------------------
+
+#ifndef NTA_BACKTRACKING_TMREGION_HPP
+#define NTA_BACKTRACKING_TMREGION_HPP
+
+#include <memory>
+#include <nupic/engine/RegionImpl.hpp>
+#include <nupic/engine/Spec.hpp>
+#include <nupic/ntypes/Value.hpp>
+#include <nupic/algorithms/BacktrackingTMCpp.hpp>
+//----------------------------------------------------------------------
+
+namespace nupic {
+class BacktrackingTMRegion : public RegionImpl {
+public:
+  typedef void (*computeCallbackFunc)(const std::string &);
+  typedef std::map<std::string, Spec> SpecMap;
+
+  BacktrackingTMRegion(const ValueMap &params, Region *region);
+  BacktrackingTMRegion(BundleIO &bundle, Region *region);
+  virtual ~BacktrackingTMRegion() {}
+
+  /* -----------  Required RegionImpl Interface methods ------- */
+
+  // Used by RegionImplFactory to create and cache
+  // a nodespec. Ownership is transferred to the caller.
+  static Spec *createSpec();
+
+  std::string getNodeType() { return "BacktrackingTMRegion"; };
+
+  // Compute outputs from inputs and internal state
+  void compute() override;
+  std::string executeCommand(const std::vector<std::string> &args,
+                             Int64 index) override;
+
+  /**
+   * Inputs/Outputs are made available in initialize()
+   * It is always called after the constructor (or load from serialized state)
+   */
+  void initialize() override;
+
+  void serialize(BundleIO &bundle) override;
+  void deserialize(BundleIO &bundle) override;
```

Comment:
do regions have different API for serialization? If not, please stick with save/load from Serializable.

Reply:
This is dictated by the plugin API so cannot change that.   This is the instruction to the plugin to do this action.  This would then call our save and load on the algorithm to do the actual work.